packages/vendure-plugin-shipmate/src/api/shipmate.service.ts

```diff
@@ -0,0 +1,271 @@
+import { Injectable, OnModuleInit, Inject, HttpStatus } from '@nestjs/common';
+import { HttpService } from '@nestjs/axios';
+import { firstValueFrom } from 'rxjs';
+import {
+  GetTokenRespose,
+  CreateShipmentResponse,
+  EventPayload,
+  TrackingEventPayload,
+} from '../types';
+import {
+  Channel,
+  ChannelService,
+  EntityHydrator,
+  EventBus,
+  FulfillmentState,
+  Logger,
+  Order,
+  OrderPlacedEvent,
+  OrderService,
+  RequestContext,
+  TransactionalConnection,
+  UserInputError,
+  isGraphQlErrorResult,
+  manualFulfillmentHandler,
+} from '@vendure/core';
+import { PLUGIN_INIT_OPTIONS, loggerCtx } from '../constants';
+import { ShipmatePluginConfig } from '../shipmate.plugin';
+import { ShipmateConfigService } from './shipmate-config.service';
+import { parseOrder } from './util';
+import { generatePublicId } from '@vendure/core/dist/common/generate-public-id';
+import { FulfillOrderInput } from '@vendure/common/lib/generated-types';
+import { Response } from 'express';
+
+export const SHIPMATE_TOKEN_HEADER_KEY = 'X-SHIPMATE-TOKEN';
+export const SHIPMATE_API_KEY_HEADER_KEY = 'X-SHIPMATE-API-KEY';
+
+@Injectable()
+export class ShipmateService implements OnModuleInit {
+  tokens: Map<string, string> = new Map();
+  constructor(
+    private httpService: HttpService,
+    @Inject(PLUGIN_INIT_OPTIONS) private config: ShipmatePluginConfig,
+    private shipmateConfigService: ShipmateConfigService,
+    private orderService: OrderService,
+    private eventBus: EventBus,
+    private connection: TransactionalConnection,
+    private entityHydrator: EntityHydrator
+  ) {}
+
+  async onModuleInit(): Promise<void> {
+    this.httpService.axiosRef.defaults.headers.common['Accept'] =
+      'application/json';
+    this.httpService.axiosRef.defaults.headers.common['Content-Type'] =
+      'application/json';
+    this.eventBus.ofType(OrderPlacedEvent).subscribe(async ({ ctx, order }) => {
+      await this.setupRequestHeaders(ctx);
+      await this.createShipment(ctx, order);
+    });
+  }
+
+  async setupRequestHeaders(ctx: RequestContext) {
+    const shipmateConfig = await this.shipmateConfigService.getConfig(ctx);
+    if (!shipmateConfig) {
+      Logger.error(
+        `Shipmate credentials not configured for channel ${ctx.channel.code}`,
+        loggerCtx
+      );
+      return;
+    }
+    let shipmateTokenForTheCurrentChannel = this.tokens.get(
+      shipmateConfig.apiKey
+    );
+    if (!shipmateTokenForTheCurrentChannel) {
+      shipmateTokenForTheCurrentChannel = await this.getShipmentToken(
+        shipmateConfig.username,
+        shipmateConfig.password
+      );
+      this.tokens.set(shipmateConfig.apiKey, shipmateTokenForTheCurrentChannel);
+    }
+    this.httpService.axiosRef.defaults.headers.common[
+      SHIPMATE_TOKEN_HEADER_KEY
+    ] = shipmateTokenForTheCurrentChannel;
+    this.httpService.axiosRef.defaults.headers.common[
+      SHIPMATE_API_KEY_HEADER_KEY
+    ] = shipmateConfig.apiKey;
```

Comment:
This is not concurrency-safe: When 2 channels simultatiously do a shipmate action, both will set the headers, but they set headers on the same `this.httpService.axiosRef` instance, so only the last one will be persisted. This means that the first channel will be making calls with the headers of the second channel. This would create orders of channel 1 in the shipmate account of channel 2. 

A good practice is to get a client ( or axios instance) on every action, based on the CTX. Similar to how we do that in the accept-blue pllugin.

1. Axios instance example (pseudo code)
```ts
getClient(ctx: RequestContext): Promise<AxiosInstance> {
   const client = axios.create();
   client.setHeaders.... etc
}
```
2. Even better: Create a separate client class that abstracts the API calls, and does HTTP validations etc.
```ts
// separate client class
export class ShipmateClient {
    constructor(username, password) {
        // create client with username password
        }

        // pass in the already transformed shipment object ready for Shipmate API
        createShipment(shipment) {
            // Get API token based on username password
            // Do the create call
            // Validate response and throw error if it didnt work
        }
}

// In this service we can then do:
getClient(ctx: RequestContext): Promise<ShipmateClient> {
    const config = getConfig(ctx);
    return new ShipmateClient(config.username, config.password);
 }
```

Reply:
Since creating the axios instance depends on the credentials provided by `ShipmateConfigService.getConfig`, it will only be encapsulating `post` calls. Do you think this is worth having its own class @martijnvdbrug ?

Reply:
I guess we can make the `ShipmateClient` it self a service class